### PR TITLE
90% AWS lambda

### PR DESCRIPTION
**Pull request by rgubby**
Created an AWS lambda compatible version of the citeproc server.

Used https://github.com/Tim-B/grunt-aws-lambda to test/create/deploy the lambda function. I've then created an API gateway resource that points to the Lambda.

``` bash
$ curl -v -X POST "https://nx5z8j2bme.execute-api.eu-west-1.amazonaws.com/staging?responseformat=json&style=harvard-cite-them-right"
{"bibliography":[{"maxoffset":0,"entryspacing":1,"linespacing":1,"second-field-align":false,"entry_ids":[],"bibliography_errors":[],"done":false,"bibstart":"<blockquote class=\"csl-bib-body\">\n","bibend":"</blockquote>"},[]]}
```

Citeproc is a bit of a mess. I've tried to refactor the `zcite` and `zcreq` objects into something more central, but I start going down a rabbit hole that I feel I am never going to escape from. I have deprecated `./lib/citeServer.js` and added a note to explain why this is the case.

In order to run this locally, i've added an http wrapper (`api-gateway.js`). This fires up a local server, creates a payload that looks the same as API gateway sends to Lambda (with our mapping template applied). This means you can make curl requests to it and test/develop it locally without having to rely on any external services (Lambda/API Gateway) running.
#### To deploy this to production

Build:
- The `grunt dist` task will create a distribution zip in the dist folder
- This zip will be copied to s3

Deploy:
- A hercules task to be created (using ansible lambda under the hood), which tells AWS Lambda to update the function from an s3 location
#### Todo
- [x] The new citeprocServerLambda.js has lots of duplicated code. We should look into extracting this duplicated code into a separate object and then requiring it from the old citeproc server and the new one.
- [x] Tests for this (via lambda_invoke maybe?)
- [ ] Update the build to package up and deploy the lambda function automatically.
- [x] Consolidate all the logging - there is far too much noise and it's clogging up the CloudWatch logs.
- [x] The zip file is 8mb whereas the .tar.gz version is bigger - check why this is the case. This is because of a load of node_modules installed but not used.

### CODE REVIEW

**Comment by rgubby** [comment]
Resolves https://github.com/talis/infra/issues/42 - no more load balancer, no more nginx.

**Comment by rgubby** [comment]
Resolves #8

**Comment by rgubby** [comment]
Resolves https://github.com/talis/infra/issues/42

File: README.md [hunk 1]
```diff
@@ -2,21 +2,37 @@
 
 ## install
 
-> git clone --recursive git@github.com:talis/citeproc-node.git
+```bash
+$ git clone --recursive git@github.com:talis/citeproc-node.git
+$ npm rebuild // rebuild step probably isn't necessary, but won't hurt
```
Comment (kiyanwang) [inline README.md]: if the version of node is fixed and you know it works against that then this is probably overkill (and slower) only really makes sense to do this if the version of node might be changing.

File: Gruntfile.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    console.log('Testing style:'+style);
+
+    grunt.task.run('createEvent:'+style);
+    grunt.task.run('lambda_invoke:'+style);
+  });
+
+  grunt.registerTask('test', function(){
+    grunt.task.run(['testStyles', 'clean:events']);
+  });
+  grunt.registerTask('dist', ['clean:dist', 'clean:events', 'lambda_package', 'lambda_deploy']);
```
Comment (kiyanwang) [inline Gruntfile.js]: i think its worth separating `dist` which should really be about creating a `distribution` from `deploy` which should be about deploying a distribution.

File: api-gateway.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      this.response.writeHead(200);
+      this.response.end(data);
+    },
+    fail:function(err){
+      this.response.writeHead(400);
+      this.response.end(err);
+    }
+  };
+
+var PORT=8181;
```
Comment (kiyanwang) [inline api-gateway.js]: should this be configurable?

**Comment by kiyanwang** [comment]
thats looking ok, just a couple of minor things noted.

**Comment by kiyanwang** [comment]
:+1:

**Comment by scaleupcto** [comment]
Should this really be in its own project? citeproc-lambda?

**Comment by rgubby** [comment]
The new code will only ever work on Lambda - so I could understand renaming it. I'm ok with that if we want to move it.

> Should this really be in its own project? citeproc-lambda?

**Comment by scaleupcto** [comment]
Rich and I talked about a much simpler approach in it's own repo built off https://github.com/michaelmcmillan/citeproc-node-iojs.

My issue with this approach is we are bending citeproc-node to do something it was never intended to do. Also we carry the baggage of the http server and god knows what else with us if we extend our fork.

My preference is for a simpler, new public repo which takes just the core citeproc-js (wrapped into an npm by the above linked project) and just a dozen lines of code to wrap that as a lambda function.

**Comment by rgubby** [comment]
New repo with this functionality in: https://github.com/talis/citeproc-lambda

**Pull request closed without merging**